swarm/storage/ldbstore.go

```diff
@@ -247,55 +308,71 @@ func decodeData(addr Address, data []byte) (*chunk, error) {
 	return NewChunk(addr, data[32:]), nil
 }
 
-func (s *LDBStore) collectGarbage(ratio float32) {
-	log.Trace("collectGarbage", "ratio", ratio)
-
-	metrics.GetOrRegisterCounter("ldbstore.collectgarbage", nil).Inc(1)
+func (s *LDBStore) collectGarbage() error {
 
-	it := s.db.NewIterator()
-	defer it.Release()
+	// the running param prevents duplicate gc from starting when one is already running
+	s.lock.Lock()
```

Comment:
And who pops the channel when we're not testing? Wouldn't that have be done within the `collectGarbage` method, after the gc is done.

However, since we might pop this in the test, then we then need to check the `default` case in the `collectGarbage` to not get stuck. The test would be the only place we could pop this externally.

If so, that means the semantics of the `collectGarbage` is dependent on the test, which seems confusing to me.

Reply:
Ok, the other way around. That will work, I guess. THanks.

Reply:
It's elegant, this solution. Good thinking.